### ISSUE TITLE
 failed to save outputs: fork/exec /usr/local/bin/docker: exec format error

### ISSUE THREAD

**Issue by xybuty** [kubeflow/pipelines#5347]
### What steps did you take:
[A clear and concise description of what the bug is.]
**I upload a pipeline to load data,the function's code like this**
![image](https://user-images.githubusercontent.com/44302306/111936955-b605e580-8b01-11eb-9d30-e1f306cd0c65.png)
**Dockerfile**
![image](https://user-images.githubusercontent.com/44302306/111937036-e8afde00-8b01-11eb-8457-8676e254dd18.png)
And the pipeline's code like this
![image](https://user-images.githubusercontent.com/44302306/111937087-ffeecb80-8b01-11eb-9a2b-55083dd64d0f.png)

### What happened:
The load-data pipeline was uploaded successful,but when I create a run based on the load-data pipeline,it occur this error
![image](https://user-images.githubusercontent.com/44302306/111937325-7390d880-8b02-11eb-9c05-526063c72c5e.png)

### Environment:

My environment is
docker version :1.13.1
k8s : v1.19.4
kubeflow:1.0.2
pipelines:0.2.5

How did you deploy Kubeflow Pipelines (KFP)?

KFP version: 
KFP SDK version: 
![image](https://user-images.githubusercontent.com/44302306/111937500-d1bdbb80-8b02-11eb-8e4d-b77bb5a88adf.png)

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

**Comment by Ark-kun** [comment]
I think the main reason for the error is that you're not using full paths in `file_outputs`. The system cannot find those files.
Can you try specifying the full paths everywhere?

Additionally I see couple of other issues (although they should not be leading to errors).

* Directly constructing `ContainerOp` objects is deprecated. Please create components in `component.yaml` format. Its very easy and is supported. See https://github.com/kubeflow/pipelines/blob/master/samples/tutorials/Data%20passing%20in%20python%20components.ipynb and https://www.kubeflow.org/docs/pipelines/sdk/component-development/
* We encourage users to specify the program command in the component definition (and not just relying on the container ENTRYPOINT)

**Comment by Ark-kun** [comment]
I've tried to recreate your case using the best practices:
```python
# Component:
from kfp.components import create_component_from_func, inputPath, OutputPath

def sklearn_datasets_load_boston(
    test_size: float = 0.33,
    x_train_path: OutputPath('NumPy'),
    y_train_path: OutputPath('NumPy'),
    x_test_path: OutputPath('NumPy'),
    y_test_path: OutputPath('NumPy'),
): 
    import os
    import numpy as np
    from sklearn import datasets
    from sklearn.model_selection import train test_split 
    x, y = datasets.load_boston(return_X_y=True)
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=test_size)
    np.save(x_train_path, x_train)
    np.save(y_train_path, y_train)
    np.save(x_test_path, x_test)
    np.save(y_test_path, y_test)


if __name__ == '__main__':
    sklearn_datasets_load_boston_op = create_component_from_func(
        sklearn_datasets_load_boston,
        packages_to_install=['scikit-learn==0.24.1'],
        output_component_file='component.yaml',
    )


# Pipeline:
def load_data_pipeline():
    load_data_task = sklearn_datasets_load_boston_op ()


if __name__ == '__main__':
    import kfp
    kfp_endpoint = None
    kfp.Client(host=kfp_endpoint).create_run_from_pipeline_func(
        load_data_pipeline,
        arguments={}
    )
```

**Comment by Ark-kun** [comment]
I've tried to recreate your case using the best practices:

```python
# Component:
from kfp.components import create_component_from_func, InputPath, OutputPath

def sklearn_datasets_load_boston(
    x_train_path: OutputPath('NumPy'),
    y_train_path: OutputPath('NumPy'),
    x_test_path: OutputPath('NumPy'),
    y_test_path: OutputPath('NumPy'),
    test_size: float = 0.33,
): 
    import os
    import numpy as np
    from sklearn import datasets
    from sklearn.model_selection import train_test_split 

    x, y = datasets.load_boston(return_X_y=True)
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=test_size)

    with open(x_train_path, 'wb') as x_train_file:
        np.save(x_train_file, x_train)
    with open(y_train_path, 'wb') as y_train_file:
        np.save(y_train_file, y_train)
    with open(x_test_path, 'wb') as x_test_file:
        np.save(x_test_file, x_test)
    with open(y_test_path, 'wb') as y_test_file:
        np.save(y_test_file, y_test)
    


if __name__ == '__main__':
    sklearn_datasets_load_boston_op = create_component_from_func(
        sklearn_datasets_load_boston,
        packages_to_install=['scikit-learn==0.24.1'],
        output_component_file='component.yaml',
    )


# Pipeline:
def load_data_pipeline():
    load_data_task = sklearn_datasets_load_boston_op(test_size=0.33)


if __name__ == '__main__':
    import kfp
    kfp_endpoint = None
    kfp.Client(host=kfp_endpoint).create_run_from_pipeline_func(
        load_data_pipeline,
        arguments={}
    )
```

**Comment by xybuty** [comment]
> I've tried to recreate your case using the best practices:
> 
> ```python
> # Component:
> from kfp.components import create_component_from_func, InputPath, OutputPath
> 
> def sklearn_datasets_load_boston(
>     x_train_path: OutputPath('NumPy'),
>     y_train_path: OutputPath('NumPy'),
>     x_test_path: OutputPath('NumPy'),
>     y_test_path: OutputPath('NumPy'),
>     test_size: float = 0.33,
> ): 
>     import os
>     import numpy as np
>     from sklearn import datasets
>     from sklearn.model_selection import train_test_split 
> 
>     x, y = datasets.load_boston(return_X_y=True)
>     x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=test_size)
> 
>     with open(x_train_path, 'wb') as x_train_file:
>         np.save(x_train_file, x_train)
>     with open(y_train_path, 'wb') as y_train_file:
>         np.save(y_train_file, y_train)
>     with open(x_test_path, 'wb') as x_test_file:
>         np.save(x_test_file, x_test)
>     with open(y_test_path, 'wb') as y_test_file:
>         np.save(y_test_file, y_test)
>     
> 
> 
> if __name__ == '__main__':
>     sklearn_datasets_load_boston_op = create_component_from_func(
>         sklearn_datasets_load_boston,
>         packages_to_install=['scikit-learn==0.24.1'],
>         output_component_file='component.yaml',
>     )
> 
> 
> # Pipeline:
> def load_data_pipeline():
>     load_data_task = sklearn_datasets_load_boston_op(test_size=0.33)
> 
> 
> if __name__ == '__main__':
>     import kfp
>     kfp_endpoint = None
>     kfp.Client(host=kfp_endpoint).create_run_from_pipeline_func(
>         load_data_pipeline,
>         arguments={}
>     )
> ```

**Comment by xybuty** [comment]
> I've tried to recreate your case using the best practices:
> 
> ```python
> # Component:
> from kfp.components import create_component_from_func, InputPath, OutputPath
> 
> def sklearn_datasets_load_boston(
>     x_train_path: OutputPath('NumPy'),
>     y_train_path: OutputPath('NumPy'),
>     x_test_path: OutputPath('NumPy'),
>     y_test_path: OutputPath('NumPy'),
>     test_size: float = 0.33,
> ): 
>     import os
>     import numpy as np
>     from sklearn import datasets
>     from sklearn.model_selection import train_test_split 
> 
>     x, y = datasets.load_boston(return_X_y=True)
>     x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=test_size)
> 
>     with open(x_train_path, 'wb') as x_train_file:
>         np.save(x_train_file, x_train)
>     with open(y_train_path, 'wb') as y_train_file:
>         np.save(y_train_file, y_train)
>     with open(x_test_path, 'wb') as x_test_file:
>         np.save(x_test_file, x_test)
>     with open(y_test_path, 'wb') as y_test_file:
>         np.save(y_test_file, y_test)
>     
> 
> 
> if __name__ == '__main__':
>     sklearn_datasets_load_boston_op = create_component_from_func(
>         sklearn_datasets_load_boston,
>         packages_to_install=['scikit-learn==0.24.1'],
>         output_component_file='component.yaml',
>     )
> 
> 
> # Pipeline:
> def load_data_pipeline():
>     load_data_task = sklearn_datasets_load_boston_op(test_size=0.33)
> 
> 
> if __name__ == '__main__':
>     import kfp
>     kfp_endpoint = None
>     kfp.Client(host=kfp_endpoint).create_run_from_pipeline_func(
>         load_data_pipeline,
>         arguments={}
>     )
> ```

Thank you so much for your reply,I‘m trying to figure out the difference between our code.
**And I have another question:how to point  kfp_endpoint?**
My KF-cluster is deployed on two local virtual machines,I wrote code in the local python environment and upload it to the cluster through the UI interface;How should I point  the kfp_endpoint?
The master node'ip is 192.168.52.10,the svc state like these
![image](https://user-images.githubusercontent.com/44302306/112102703-a363dd00-8be3-11eb-9458-a63a6b0dc166.png)

**Comment by Ark-kun** [comment]
>I‘m trying to figure out the difference between our code.

The root issue in your code was probably using non-absolute paths in `file_outputs`.

While fixing the issue I've also discovered that `np.save(data_path, data)` writes to `data_path + ".npy"`, which was leading to a similar error message in my code. To fix that I've used `np.save` with pre-opened files.

**Comment by Ark-kun** [comment]
>how to point kfp_endpoint?
>My KF-cluster is deployed on two local virtual machines,I wrote code in the local python environment and upload it to the cluster through the UI interface;How should I point the kfp_endpoint?

I'm not sure the direct connection to the API serve with `kfp_endpoint = "http://10.100.203.15:8888"` will work.
You'll probably need to port forward the api server `kubectl port-forward svc/ml-pipeline 8888` and use `kfp_endpoint 
 = http://127.0.0.1:8888`

/cc @Bobgy

**Comment by xybuty** [comment]
> > I‘m trying to figure out the difference between our code.
> 
> The root issue in your code was probably using non-absolute paths in `file_outputs`.
> 
> While fixing the issue I've also discovered that `np.save(data_path, data)` writes to `data_path + ".npy"`, which was leading to a similar error message in my code. To fix that I've used `np.save` with pre-opened files.

thank you so much for your reply ,After I improved my code according to your guidance,I found maybe my problem is not same as this problem [https://github.com/kubeflow/pipelines/issues/750].Because the load_data image can save data to the path in containter,like this
![image](https://user-images.githubusercontent.com/44302306/112306006-4a2aa500-8cda-11eb-8d08-715399950994.png)
Then I  check the pod logs I get the following error.
![image](https://user-images.githubusercontent.com/44302306/112306332-aa214b80-8cda-11eb-95b3-d8d15ae999d0.png)
Look forward your help!

**Comment by xybuty** [comment]
> > I‘m trying to figure out the difference between our code.
> 
> The root issue in your code was probably using non-absolute paths in `file_outputs`.
> 
> While fixing the issue I've also discovered that `np.save(data_path, data)` writes to `data_path + ".npy"`, which was leading to a similar error message in my code. To fix that I've used `np.save` with pre-opened files.
The screenshot is not clear,here is the information of the error
`time="2021-03-24T19:04:42Z" level=info msg="Creating a docker executor"
time="2021-03-24T19:04:42Z" level=info msg="Executor (version: v2.3.0, build_date: 2019-05-20T22:10:54Z) initialized (pod: kubeflow/load-data-2-mctsc-3737213461) with template:\n{\"name\":\"load-data\",\"inputs\":{},\"outputs\":{\"artifacts\":[{\"name\":\"load-data-X_train\",\"path\":\"/X_train.npy\"}]},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"load_data2:v0.0.2\",\"command\":[\"python\",\"load_data.py\"],\"resources\":{}},\"archiveLocation\":{\"s3\":{\"endpoint\":\"minio-service.kubeflow:9000\",\"bucket\":\"mlpipeline\",\"insecure\":true,\"accessKeySecret\":{\"name\":\"mlpipeline-minio-artifact\",\"key\":\"accesskey\"},\"secretKeySecret\":{\"name\":\"mlpipeline-minio-artifact\",\"key\":\"secretkey\"},\"key\":\"artifacts/load-data-2-mctsc/load-data-2-mctsc-3737213461\"}}}"
time="2021-03-24T19:04:42Z" level=info msg="Waiting on main container"
time="2021-03-24T19:04:42Z" level=info msg="main container started with container ID: 9f20576bc2d9b749429bc259cf10c932a4636fbdd1041043c8cc5550d2d74158"
time="2021-03-24T19:04:42Z" level=info msg="Starting annotations monitor"
time="2021-03-24T19:04:42Z" level=info msg="docker wait 9f20576bc2d9b749429bc259cf10c932a4636fbdd1041043c8cc5550d2d74158"
time="2021-03-24T19:04:42Z" level=info msg="Starting deadline monitor"
time="2021-03-24T19:04:42Z" level=error msg="executor error: fork/exec /usr/local/bin/docker: exec format error\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/common.RunCommand\n\t/go/src/github.com/argoproj/argo/workflow/common/util.go:397\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:95\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:867\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:32\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2021-03-24T19:04:42Z" level=info msg="No sidecars"
time="2021-03-24T19:04:42Z" level=info msg="No output parameters"
time="2021-03-24T19:04:42Z" level=info msg="Saving output artifacts"
time="2021-03-24T19:04:42Z" level=info msg="Staging artifact: load-data-X_train"
time="2021-03-24T19:04:42Z" level=info msg="Copying /X_train.npy from container base image layer to /argo/outputs/artifacts/load-data-X_train.tgz"
time="2021-03-24T19:04:42Z" level=info msg="Archiving 9f20576bc2d9b749429bc259cf10c932a4636fbdd1041043c8cc5550d2d74158:/X_train.npy to /argo/outputs/artifacts/load-data-X_train.tgz"
time="2021-03-24T19:04:42Z" level=info msg="sh -c docker cp -a 9f20576bc2d9b749429bc259cf10c932a4636fbdd1041043c8cc5550d2d74158:/X_train.npy - | gzip > /argo/outputs/artifacts/load-data-X_train.tgz"
time="2021-03-24T19:04:42Z" level=info msg="Annotations monitor stopped"
time="2021-03-24T19:04:42Z" level=warning msg="path /X_train.npy does not exist (or /X_train.npy is empty) in archive /argo/outputs/artifacts/load-data-X_train.tgz"
time="2021-03-24T19:04:42Z" level=error msg="executor error: path /X_train.npy does not exist (or /X_train.npy is empty) in archive /argo/outputs/artifacts/load-data-X_train.tgz\ngithub.com/argoproj/argo/errors.New\n\t/go/src/github.com/argoproj/argo/errors/errors.go:49\ngithub.com/argoproj/argo/errors.Errorf\n\t/go/src/github.com/argoproj/argo/errors/errors.go:55\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).CopyFile\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:66\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).stageArchiveFile\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:344\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).saveArtifact\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:245\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).SaveArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:231\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:54\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2021-03-24T19:04:42Z" level=info msg="Alloc=4108 TotalAlloc=11155 Sys=70590 NumGC=4 Goroutines=8"
time="2021-03-24T19:04:42Z" level=fatal msg="path /X_train.npy does not exist (or /X_train.npy is empty) in archive /argo/outputs/artifacts/load-data-X_train.tgz\ngithub.com/argoproj/argo/errors.New\n\t/go/src/github.com/argoproj/argo/errors/errors.go:49\ngithub.com/argoproj/argo/errors.Errorf\n\t/go/src/github.com/argoproj/argo/errors/errors.go:55\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).CopyFile\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:66\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).stageArchiveFile\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:344\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).saveArtifact\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:245\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).SaveArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:231\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:54\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"`

**Comment by Ark-kun** [comment]
Do you still get the error when you run the pipeline code I've posted?

**Comment by xybuty** [comment]
> Do you still get the error when you run the pipeline code I've posted? It worked for me.

YES，I still got the error (T_T)

**Comment by Bobgy** [comment]
Duplicate of https://github.com/kubeflow/pipelines/issues/1654

I think this is the problem -- kubernetes 1.19  has dropped built-in docker support. So you should switch to use argo pns executor.

**Comment by xybuty** [comment]
> Duplicate of #1654
> 
> I think this is the problem -- kubernetes 1.19 has dropped built-in docker support. So you should switch to use argo pns executor.
YES,You are right,this problem didn't reappear after Ichange rhe k8s version。

**Comment by Ark-kun** [comment]
Looks like this particular issue is resolved.
